src/core/stdcpp/allocator.d

```diff
@@ -0,0 +1,50 @@
+/**
+ * D header file for interaction with C++ std::allocator.
+ *
+ * Copyright: Copyright (c) 2018 D Language Foundation
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Guillaume Chatelet
+ *            Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/allocator.d)
+ */
+
+module core.stdcpp.allocator;
+
+extern(C++, "std"):
+
+/**
+ * Allocators are classes that define memory models to be used by some parts of
+ * the C++ Standard Library, and most specifically, by STL containers.
+ */
+extern(C++, class) struct allocator(T)
+{
+    static assert(!is(T == const), "The C++ Standard forbids containers of const elements because allocator!(const T) is ill-formed.");
```

Comment:
Should also check for `immutable` no ?

Reply:
Sure. What I really care about though, is working out how to run those tests in this pr to prove it works on all those build machines...

Reply:
I'm pretty sure it does not work on POSIX, because [template mangling is broken](https://github.com/dlang/dmd/pull/8455).
But having those tests running on Windows would be awesome.

Reply:
We need to run it on every target, then I'd know that template mangling is broken and fix it...
I'm not aware of problems with template mangling. This needs to work everywhere. Can't know that unless ci executes the tests.

Reply:
`immutable` should already be caught by the mangling code, because there is no equivalent.

Reply:
I don't actually think there's any reason to believe these C++ containers will exclusively interact with `extern(C++)`... once you interface C++ and these are floating around, they'll probably be used generally by D code too.

Reply:
That's right, however the additional check is a small price to pay here. A casual reader might not realize that `immutable` is disallowed by the compiler. Also, we have no guarantee this won't change in the future. Lastly, it will provide a better error message.

Reply:
These are all great ideas, but it's all worthless unless we find a way to test it ;)

Reply:
See https://github.com/TurkeyMan/druntime/pull/1. I suspect it will conflict with master due to other added tests, though.

Reply:
@rainers Thanks! I can't see how the `.cpp` file gets built in there though, that's the trouble point, since it needs to compile for all targets, and that includes DMC and MSVC on windows...

Reply:
> I can't see how the `.cpp` file gets built in there though, that's the trouble point, since it needs to compile for all targets, and that includes DMC and MSVC on windows...

The amended changes to the makefile now contain the cpp-compilation, but as string.d doesn't compile on anything but win64, the auto-tester stops before it schedules anything to the Windows machines. I'd suggest adding `version(CRuntime_Microsoft):` at the top for now.

When this works to some extend, adding makefiles for other targets to the test should be straight-forward.